### PR TITLE
restrict certificate type for builtin SSH server

### PR DESCRIPTION
**Pull request by earl-warren**
- While doing some sanity checks over OpenSSH's code for how they handle certificates authentication. I stumbled on an condition that checks the certificate type is really an user certificate on the server-side authentication. This checks seems to be a formality and just for the sake of good domain seperation, because an user and host certificate don't differ in their generation, verification or flags that can be included.
- Add this check to the builtin SSH server to stay close to the unwritten SSH specification.
- This is an breaking change for setups where the builtin SSH server is being used and for some reason host certificates were being used for authentication.

(cherry picked from commit de35b141b79a3d6efe2127ed2c73fd481515e481)

Refs: https://codeberg.org/forgejo/forgejo/pulls/1172

## :warning: BREAKING :warning:

The built-in SSH server will now only accept SSH user certificates, not server certificates. This behaviour matches OpenSSH.

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

File: modules/ssh/ssh.go [hunk 1]
```diff
@@ -191,6 +191,12 @@ func publicKeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
 			return false
 		}
 
+		if cert.CertType != gossh.UserCert {
+			log.Warn("Certificate Rejected: Not a user certificate")
+			log.Warn("Failed authentication attempt from %s", ctx.RemoteAddr())
```
Comment (puni9869) [inline modules/ssh/ssh.go:196]: could you merge these in one.

**Review by puni9869 — APPROVED** [review]

File: modules/ssh/ssh.go (hunk 1 above)
Comment (silverwind) [reply modules/ssh/ssh.go:196]: Think it's fine. Go is not really made for multi-line strings.

**Comment by silverwind** [comment]
Breaking label added, good to go imho.

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
In the unlikely event this is breaking someone's workflow, I guess a option could be introduced to skip this check, but as I see it, this is [not configurable](https://man.openbsd.org/sshd_config) in OpenSSH either.

**Pull request merged**